org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java

```diff
@@ -0,0 +1,1766 @@
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.gecko.emf.exporter.cells.EMFExportEObjectIDValueCell;
+import org.gecko.emf.exporter.cells.EMFExportEObjectManyReferencesValueCell;
+import org.gecko.emf.exporter.cells.EMFExportEObjectOneReferenceValueCell;
+import org.gecko.emf.exporter.cells.EMFExportMappingMatrixReferenceValueCell;
+import org.gecko.emf.exporter.headers.EMFExportEObjectColumnHeader;
+import org.gecko.emf.exporter.headers.EMFExportEObjectGenericColumnHeader;
+import org.gecko.emf.exporter.headers.EMFExportEObjectIDColumnHeader;
+import org.gecko.emf.exporter.headers.EMFExportEObjectManyReferencesColumnHeader;
+import org.gecko.emf.exporter.headers.EMFExportEObjectOneReferenceColumnHeader;
+import org.gecko.emf.exporter.keys.EMFExportRefMatrixNameIDCompositeKey;
+import org.slf4j.Logger;
+
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+
+/**
+ * Implementation of format-agnostic methods for exporting EMF resources and
+ * lists of EMF objects to desired format.
+ * 
+ * Format-specific implementations should extend this class and only focus on
+ * transformation of matrices resulting from {@link #exportEObjectsToMatrices}
+ * to arrive at desired format.
+ * 
+ * @author Michal H. Siemaszko
+ */
+public abstract class AbstractEMFExporter implements EMFExporter {
+	protected static final List<String> METADATA_ECLASS_MATRIX_COLUMNS_HEADERS = List.of("Name", "Type", "isMany",
+			"isRequired", "Default value", "Documentation");
+
+	protected static final List<String> METADATA_EENUM_MATRIX_COLUMNS_HEADERS = List.of("Name", "Literal", "Value",
+			"Documentation");
+
+	protected static final String METADATA_DOCUMENTATION_HEADER = "Documentation";
+	protected static final String METADATA_PSEUDOID_DOCUMENTATION = "No default ID present, therefore pseudo-ID was generated for identification purposes";
+
+	protected static final String METADATA_MATRIX_NAME_SUFFIX = "Metadata";
+	protected static final String MAPPING_MATRIX_NAME_SUFFIX = "Mapping";
+
+	protected static final String DOCUMENTATION_GENMODEL_SOURCE = "http://www.eclipse.org/emf/2002/GenModel";
+	protected static final String DOCUMENTATION_GENMODEL_DETAILS = "documentation";
+
+	protected static final String ECORE_PACKAGE_NAME = "ecore";
+
+	protected static final String ID_COLUMN_NAME = "Id";
+
+	protected static final String ID_ATTRIBUTE_NAME = "id";
+
+	protected static final String REF_COLUMN_PREFIX = "ref_";
+
+	protected static final int MAX_COLUMNS = 16384;
+
+	protected static final int MAX_ROWS = 1048576;
+
+	private final Logger logger;
+
+	protected final Stopwatch stopwatch;
+
+	// maps EObjects' IDs to names of matrices, so those can be looked up e.g. when
+	// constructing links
+	protected final Map<String, String> eObjectIDToMatrixNameMap;
```

Comment:
We must make the services, that result from this abstract class, stateless and hand the lists and maps used for caching everywhere we go (maybe via a Context, to not explode the method signatures). 

Even though the services I saw are prototyped, this is a Problem. For one, the user of the Services does not know that the service is stateful and thus, does not necessary use a new instance for eache usage. This creates memory leaks and also may produce concurrency issues.

This already happens e.g. in the `EMFXLSXResourceFactoryConfigurator`, which uses one instance for as long as it exists.

Reply:
All these data structures are reset via `org.gecko.emf.exporter.AbstractEMFExporter.resetState()` with each call to public methods.

As you know, these data structures are needed to speed up (hundreds of times speed up) processing, to be able to construct lookup indexes / references / links / etc.
I moved them to instance level as there are already 6 such data structures in `org.gecko.emf.exporter.AbstractEMFExporter` and even more in format-specific exporters (i.e. additional 4 in `org.gecko.emf.exporter.csv.impl.EMFCSVExporter`, etc.).
 
Is resetting state on each call to public method (via `org.gecko.emf.exporter.AbstractEMFExporter.resetState()`) an acceptable solution for you or should I proceed with refactor and introduce a parameter object where these data structures are held ?

Reply:
Not sure if this would cut it. I would Expect that the Exporter can handle concurrent calls to the export method on the same instance. I believe that even a resetState method will not make it Thread safe.

Reply:
I will introduce a parameter object then. 

I was under the impression, that since a lot of EMF and Gecko stuff is not thread safe, and here I'm using scope `PROTOTYPE`, this will be sufficient. But I will change this as you prefer then.